File: hitch/blueprints/utils/ride_score.py
"""Python mirror of hitch/static/ride_score.js. Reads the SAME canonical weights
file so the two can never disagree on point values. Used by later phases to compute
per-user aggregates from stored ride content."""

import json
import pathlib

# hitch/blueprints/utils/ride_score.py -> parents[2] == hitch/ ; weights live in hitch/static/.
_WEIGHTS_PATH = pathlib.Path(__file__).resolve().parents[2] / "static" / "ride_score_weights.json"
WEIGHTS = json.loads(_WEIGHTS_PATH.read_text())
PASSENGER_KINDS = set(WEIGHTS["passenger_kinds"])


def _is_filled(field: str, value) -> bool:
    # Mirrors isFilled() in ride_score.js.
    if isinstance(value, (list, tuple)):
        return len(value) > 0
    if field == "driver_age":
        return value is not None and str(value).strip() != ""
    return isinstance(value, str) and value.strip() != ""


def _score_group(fields: dict, weight_map: dict):
    earned = 0
    max_pts = 0
    missing = []
    for field, pts in weight_map.items():
        max_pts += pts
        if _is_filled(field, fields.get(field)):
            earned += pts
        else:
            missing.append({"field": field, "pts": pts})
    return earned, max_pts, missing


def score_fields(fields: dict) -> dict:
    fields = fields or {}
    d_earned, d_max, d_missing = _score_group(fields, WEIGHTS["driver"])
    d_pct = round(d_earned / d_max * 100) if d_max else 0

    b_earned, b_max, b_missing = _score_group(fields, WEIGHTS["vehicle_base"])
    bonus_eligible = fields.get("vehicle_kind") in PASSENGER_KINDS
    v_earned, v_max, v_missing = b_earned, b_max, list(b_missing)
    if bonus_eligible:
        x_earned, x_max, x_missing = _score_group(fields, WEIGHTS["vehicle_bonus"])
        v_earned += x_earned
        v_max += x_max
        v_missing.extend(x_missing)
    v_missing.sort(key=lambda m: m["pts"], reverse=True)
    d_missing.sort(key=lambda m: m["pts"], reverse=True)
    v_pct = round(v_earned / v_max * 100) if v_max else 0

    # Combined completeness over the whole (driver + vehicle) pool — the UI shows only this;
    # driver/vehicle stay for backend aggregates.
    max_total = d_max + v_max
    total = d_earned + v_earned
    pct = round(total / max_total * 100) if max_total else 0

    return {
        "driver": {"earned": d_earned, "max": d_max, "pct": d_pct, "missing": d_missing},
        "vehicle": {"earned": v_earned, "max": v_max, "pct": v_pct, "missing": v_missing, "bonus_eligible": bonus_eligible},
        "total": total,
        "max_total": max_total,
        "pct": pct,
    }

File: hitch/static/ride_score.js
// Pure completeness-scoring library shared by the browser (window.RideScore) and
// Node tests (module.exports). Weights are supplied by the caller from the single
// canonical source hitch/static/ride_score_weights.json — never hard-coded here.
(function (root, factory) {
  const mod = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = mod;
  else root.RideScore = mod;
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  // A field is "filled" when the user has supplied a real answer. Arrays need a
  // non-empty entry; strings need non-whitespace; age accepts a number or numeric
  // string; commercial is a tri-state where BOTH true and false count as answered
  // (only null/undefined is unanswered).
  function isFilled(field, value) {
    if (field === "commercial") return value === true || value === false;
    if (Array.isArray(value)) return value.length > 0;
    if (field === "driver_age") return value !== null && value !== undefined && String(value).trim() !== "";
    return typeof value === "string" && value.trim() !== "";
  }

  function scoreGroup(fields, weightMap) {
    let earned = 0, max = 0;
    const missing = [];
    for (const field of Object.keys(weightMap)) {
      const pts = weightMap[field];
      max += pts;
      if (isFilled(field, fields[field])) earned += pts;
      else missing.push({ field: field, pts: pts });
    }
    return { earned, max, missing };
  }

  function computeScores(fields, weights) {
    fields = fields || {};
    const driver = scoreGroup(fields, weights.driver);
    const driverPct = driver.max ? Math.round((driver.earned / driver.max) * 100) : 0;

    const base = scoreGroup(fields, weights.vehicle_base);
    const bonusEligible = weights.passenger_kinds.indexOf(fields.vehicle_kind) !== -1;
    let vEarned = base.earned, vMax = base.max;
    const vMissing = base.missing.slice();
    if (bonusEligible) {
      const bonus = scoreGroup(fields, weights.vehicle_bonus);
      vEarned += bonus.earned;
      vMax += bonus.max;
      for (const m of bonus.missing) vMissing.push(m);
    }
    // Highest-value missing first, so nudges surface the biggest wins.
    vMissing.sort((a, b) => b.pts - a.pts);
    driver.missing.sort((a, b) => b.pts - a.pts);
    const vPct = vMax ? Math.round((vEarned / vMax) * 100) : 0;

    // Combined completeness: one number over the whole (driver + vehicle) point pool.
    // The UI shows only this; driver/vehicle stay in the return for backend aggregates.
    const maxTotal = driver.max + vMax;
    const total = driver.earned + vEarned;
    const pct = maxTotal ? Math.round((total / maxTotal) * 100) : 0;

    return {
      driver: { earned: driver.earned, max: driver.max, pct: driverPct, missing: driver.missing },
      vehicle: { earned: vEarned, max: vMax, pct: vPct, missing: vMissing, bonusEligible: bonusEligible },
      total: total,
      maxTotal: maxTotal,
      pct: pct,
    };
  }

  return { computeScores };
});
